rust/private/rust.bzl

```diff
@@ -546,6 +562,15 @@ _common_attrs = {
             These are other `rust_library` targets and will be presented as the new name given.
         """),
     ),
+    # Previously `build_script` was part of `deps`, but much like proc_macro_deps we need to tell
+    # bazel that this is `cfg = "exec"` and making it its own dep line solves the problem.
+    "build_script": attr.label(
+        doc = dedent("""\
+            The `cargo_build_script` dependency, if any, to run as part of building this target.
+        """),
+        cfg = "exec",
```

Comment:
This feels incorrect to me. The build script will be executed on `exec` but it's outputs need to be for the target platform. I still think I'm confused why the current implementation is not sufficient where the `cargo_build_script` rule takes the build script itself as `cfg = "exec"` and the outputs are consumed as `cfg = "target"`.

Reply:
If you specify it as `cfg = "target"` then it gets interpreted as needing to be _executable_ on the target, which isn't true. AFAIK there's no way to specify "runs on exec but produces outputs for target" in bazel, but @scentini or @krasimirgg should confirm.

Reply:
> If you specify it as cfg = "target" then it gets interpreted as needing to be executable on the target

That's not correct. The build script is made `exec` due to https://github.com/bazelbuild/rules_rust/blob/0.10.0/cargo/cargo_build_script.bzl#L278-L284. The outputs of the build script I don't think are intended to be for the execution platform so setting this to `exec` would be incorrect in cases like [openssl-sys](https://github.com/sfackler/rust-openssl/commits/master/openssl-sys/build/main.rs) where the build script is trying to find OpenSSL libraries to link against, right?

Reply:
Last two comments both sound correct to me. Let me try to express my viewpoint to hopefully clarify something or else learn why and how it's wrong: there's a distinction between what platform the build script is running on (and built for), and what platform the outputs of the build script are targeting. Suppose we are cross-compiling on an x86 machine and we are aiming to produce final rust libraries and binaries for arm. The build script itself is ran on the x86 machine and produces outputs that are specific to the arm final target. When building the build script itself, all of its build dependencies need to be built in x86 mode in order for them to be runnable on the x86 machine (and the `cfg = exec` on the build script attribute expresses this constraint). The build script is free to output artifacts that contain instructions for any platform -- the `cfg` only controls where it's supposed to run, not what platform its outputs are supposed to be for. So yes, the outputs of the build script will contain the arm-specific definitions that will then be fed up to the rustc cross-compilation invocation (again running on x86) to produce final rlibs and bins that are for arm. In a sense, the outputs of the build script are "target-independent", but they contain target-specific instructions inside.

Reply:
I think we're on the same page but I'm still not sure I understand why this attribute should be `cfg = "exec"`. All actions consumed as `cfg = "target"` still have their actions run on some `exec` platform. But the outputs are intended to be run on the `target` platform. Is there a build script that outputs some binary that's loaded or invoked by a call to `rustc`? I otherwise still think `cfg = "target"` is correct.